### PR TITLE
*: support a region divided into multiple regions

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
Serial splitting and scattering regions take a long time.
Related to pingcap/kvproto#440, pingcap/tikv#5268

### What is changed and how it works?
If some split keys in a region, we can batch split these keys. Then we can split a region into several regions in a request. And we split of different regions in parallel.

### Check List 

Tests 
 - No code

Code changes

 - Has exported function/method change
 - Has interface methods change

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @crazycs520 @bb7133

**Comment by zimulala** [comment]
PTAL @crazycs520 @bb7133

**Comment by winkyao** [comment]
@zimulala Please give some results of performance improvement

File: ddl/split_region.go [hunk 1]
```diff
@@ -90,26 +93,27 @@ func splitPreSplitedTable(store kv.SplitableStore, tbInfo *model.TableInfo, scat
...
+	regionIDs, err := store.SplitRegions(context.Background(), [][]byte{tableStartKey}, scatter)
 	if err != nil {
 		// It will be automatically split by TiKV later.
 		logutil.BgLogger().Warn("[ddl] split table region failed", zap.Error(err))
 	}
-	return regionID
+	if len(regionIDs) == 1 {
+		return regionIDs[0]
+	}
+	return 0
```
Comment (winkyao) [inline ddl/split_region.go]: Shall we return an error here?

File: store/tikv/split_region.go [hunk 2]
```diff
@@ -16,70 +16,165 @@ package tikv
...
-				return 0, errors.Trace(err)
-			}
+	}
+	return batchResp
+}
+
+// SplitRegions splits regions contains splitKeys into some egions.
+func (s *tikvStore) SplitRegions(ctx context.Context, splitKeys [][]byte, scatter bool) (regionIDs []uint64, err error) {
+	// TODO: DO we need update this backoff?
+	bo := NewBackoffer(ctx, splitRegionBackoff)
```
Comment (winkyao) [inline store/tikv/split_region.go]: Please start a discussion about the backoff.

File: store/tikv/split_region.go [hunk 3]
```diff
@@ -16,70 +16,165 @@ package tikv
...
+		return nil, nil
+	}
+	if len(batches) == 1 {
+		resp := s.batchSendSingleRegion(bo, batches[0], scatter)
+		return resp.resp, errors.Trace(resp.err)
+	}
+	ch := make(chan singleBatchResp, len(batches))
+	for _, batch1 := range batches {
+		batch := batch1
+		go func() {
```
Comment (winkyao) [inline store/tikv/split_region.go]: util.WithRecover?

File: store/tikv/split_region.go [hunk 4]
```diff
@@ -16,70 +16,165 @@ package tikv
...
+		return batchResp
+	}
+
+	spResp := resp.Resp.(*kvrpcpb.SplitRegionResponse)
+	left := spResp.GetLeft()
+	regions := spResp.GetRegions()
+	if left != nil {
+		spResp.Regions = []*metapb.Region{left}
+	} else if len(regions) > 0 {
+		spResp.Regions = regions[:len(regions)-1]
```
Comment (crazycs520) [inline store/tikv/split_region.go]: Why use `regions[:len(regions)-1]`
Please add a comment?

File: store/tikv/split_region.go [hunk 5]
```diff
@@ -16,70 +16,165 @@ package tikv
...
+				zap.Error(batchResp.err))
+			if err == nil {
+				err = batchResp.err
+			}
+			continue
+		}
+
+		spResp := batchResp.resp.Resp.(*kvrpcpb.SplitRegionResponse)
+		regions := spResp.GetRegions()
+		srResp.Regions = append(srResp.Regions, regions[:len(regions)-1]...)
```
Comment (crazycs520) [inline store/tikv/split_region.go]: Why use `regions[:len(regions)-1]`
Please add a comment?

File: store/tikv/split_region.go [hunk 6]
```diff
@@ -16,70 +16,165 @@ package tikv
...
-		if err != nil {
-			return 0, errors.Trace(err)
+		resp, err = s.spliteBatchRegionsReq(bo, batch.keys, scatter)
+		batchResp.resp = resp
+		batchResp.err = err
+		return batchResp
+	}
+
+	spResp := resp.Resp.(*kvrpcpb.SplitRegionResponse)
+	left := spResp.GetLeft()
```
Comment (crazycs520) [inline store/tikv/split_region.go]: I see the `GetLeft` method was Deprecated.

File: ddl/split_region.go (hunk 1 above)
Comment (zimulala) [reply ddl/split_region.go]: It's the original code and several functions on the upper layer also did not return an error message.

**Comment by zimulala** [comment]
PTAL @winkyao @crazycs520 @bb7133

**Comment by zimulala** [comment]
PTAL @winkyao @nolouch

File: executor/split.go [hunk 7]
```diff
@@ -94,27 +93,18 @@ func (e *SplitIndexRegionExec) splitIndexRegion(ctx context.Context) error {
...
-				zap.String("index", e.indexInfo.Name.L),
-				zap.Error(err))
-			continue
-		}
-		if regionID == 0 {
-			continue
-		}
-		regionIDs = append(regionIDs, regionID)
-
+	regionIDs, err := s.SplitRegions(context.Background(), splitIdxKeys, true)
```
Comment (bb7133) [inline executor/split.go]: Why not using `ctxWithTimeout` here? IMHO a timeout is still needed

File: store/tikv/split_region.go [hunk 8]
```diff
@@ -16,70 +16,180 @@ package tikv
...
-		if scatter {
-			err = s.scatterRegion(left.Id)
-			if err != nil {
-				return 0, errors.Trace(err)
-			}
+	}
+	return batchResp
+}
+
+// SplitRegions splits regions contains splitKeys into some regions.
```
Comment (bb7133) [inline store/tikv/split_region.go]: The term `some regions` is confusing, how about "SplitRegions splits regions by splitKeys"?

File: store/tikv/split_region.go [hunk 9]
```diff
@@ -16,70 +16,180 @@ package tikv
...
+		resp, err = s.splitBatchRegionsReq(bo, batch.keys, scatter)
+		batchResp.resp = resp
+		batchResp.err = err
+		return batchResp
+	}
+
+	spResp := resp.Resp.(*kvrpcpb.SplitRegionResponse)
+	regions := spResp.GetRegions()
+	if len(regions) > 0 {
+		// Divide a region into n, one can not need to be scattered,
```
Comment (bb7133) [inline store/tikv/split_region.go]:
```suggestion
		// Divide a region into n, one of them may not need to be scattered,
```

File: store/tikv/split_region.go [hunk 10]
```diff
@@ -16,70 +16,180 @@ package tikv
...
+		// so n-1 needs to be scattered to other storage servers.
```
Comment (bb7133) [inline store/tikv/split_region.go]:
```suggestion
		// so n-1 needs to be scattered to other stores.
```

File: store/tikv/split_region.go [hunk 11]
```diff
@@ -16,70 +16,180 @@ package tikv
...
+		logutil.BgLogger().Info("batch split regions complete",
+			zap.Uint64("batch region ID", batch.regionID.id),
+			zap.Stringer("first at", kv.Key(batch.keys[0])),
+			zap.Stringer("first new region left", logutil.Hex(spResp.Regions[0])),
+			zap.Int("new region count", len(spResp.Regions)))
+		return batchResp
+	}
+
+	for i, r := range spResp.Regions {
+		err = s.scatterRegion(r.Id)
```
Comment (bb7133) [inline store/tikv/split_region.go]: How about `if err = s.scatterRegion(r.Id); err == nil {`?

File: store/tikv/split_region.go [hunk 12]
```diff
@@ -16,70 +16,180 @@ package tikv
...
-			if err != nil {
-				return 0, errors.Trace(err)
-			}
+	}
+	return batchResp
+}
+
+// SplitRegions splits regions contains splitKeys into some regions.
+func (s *tikvStore) SplitRegions(ctx context.Context, splitKeys [][]byte, scatter bool) (regionIDs []uint64, err error) {
+	splitRegionsBo := splitRegionBackoff * math.Min(float64(len(splitKeys)), 10)
```
Comment (bb7133) [inline store/tikv/split_region.go]: Why use `10` here?

Comment (zimulala) [reply store/tikv/split_region.go]: `splitRegionBackoff` is 20s, I think `10` is enough.

File: executor/split.go (hunk 7 above)
Comment (zimulala) [reply executor/split.go]: The original code hasn't and `SplitRegions` has itself timeout, so I don't add it.
If you feel the need, is there a suggested value?

Comment (bb7133) [reply executor/split.go]: OK, never mind

File: store/tikv/split_region.go (hunk 12 above)
Comment (bb7133) [reply store/tikv/split_region.go]: Please add some comments on this value, or use a constant.

**Pull request merged**